crates/transpiler/src/passes/disjoint_layout.rs

```diff
@@ -0,0 +1,545 @@
+        node_map.insert(node.0, new_node);
+    }
+    for edge in filtered.edge_references() {
+        let new_source = *node_map.get(&edge.source()).unwrap();
+        let new_target = *node_map.get(&edge.target()).unwrap();
+        out_graph.add_edge(new_source, new_target, ());
+    }
+    out_graph
+}
+
+#[pyfunction(name = "run_pass_over_connected_components")]
+pub fn py_run_pass_over_connected_components(
+    dag: &mut DAGCircuit,
+    target: &Target,
+    run_func: Bound<PyAny>,
+) -> PyResult<Option<Vec<PyObject>>> {
+    let func = |dag: DAGCircuit, cmap: &CouplingMap| -> PyResult<PyObject> {
+        let py = run_func.py();
+        let coupling_map_cls = COUPLING_MAP.get_bound(py);
+        let endpoints: Vec<[usize; 2]> = cmap
+            .edge_indices()
+            .map(|edge| {
+                let endpoints = cmap.edge_endpoints(edge).unwrap();
+                [endpoints.0.index(), endpoints.1.index()]
+            })
+            .collect();
+        let out_list = PyList::new(py, endpoints)?;
+        let py_cmap = coupling_map_cls.call1((out_list,))?;
+        for node in cmap.node_indices() {
+            py_cmap
+                .getattr(intern!(py, "graph"))?
+                .set_item(node.index(), cmap.node_weight(node).unwrap().index())?;
+        }
+        // TODO: Remove the dag clone
+        Ok(run_func.call1((dag, py_cmap))?.unbind())
+    };
+    match distribute_components(dag, target)? {
+        DisjointSplit::NoneNeeded => {
+            let coupling_map: CouplingMap = match build_coupling_map(target) {
+                Some(map) => map,
+                None => return Ok(None),
+            };
+            Ok(Some(vec![func(dag.clone(), &coupling_map)?]))
+        }
+        DisjointSplit::TargetSubset(qubits) => {
+            let coupling_map = build_coupling_map(target).unwrap();
+            let cmap = subgraph(
+                &coupling_map,
+                &qubits.iter().map(|x| NodeIndex::new(x.index())).collect(),
+            );
+            Ok(Some(vec![func(dag.clone(), &cmap)?]))
+        }
+        DisjointSplit::Arbitrary(components) => Some(
+            components
+                .into_iter()
+                .map(|component| {
+                    let coupling_map = build_coupling_map(target).unwrap();
+                    let cmap = subgraph(
+                        &coupling_map,
+                        &component
+                            .physical_qubits
+                            .iter()
+                            .map(|x| NodeIndex::new(x.index()))
+                            .collect(),
+                    );
+                    func(component.sub_dag, &cmap)
+                })
+                .collect::<PyResult<Vec<_>>>(),
+        )
+        .transpose(),
+    }
+}
+
+pub fn distribute_components(dag: &mut DAGCircuit, target: &Target) -> PyResult<DisjointSplit> {
+    let coupling_map: CouplingMap = match build_coupling_map(target) {
```

Comment:
We have an ordering choice here: I wrote https://github.com/Qiskit/qiskit/blob/9cbb8915fc9d1ec425082daeea1acb195d53b4c2/crates/transpiler/src/target/mod.rs#L1385-L1421 for Sabre, which has a bunch of overlap with this, but is a bit more unified and centralised to `Target`, and is more explicit about the assumptions it makes.  Perhaps we could pull that in (or I can separate it off to merge first) and add a variant that takes a `node_fn: impl FnMut(PhysicalQubit) -> T` to produce node weights?

Reply:
If you don't mind generalizing it that sounds good to me.  This also comes up in vf2 where there is a similar function already, I didn't make a generic version because this PR was racing with the vf2 one (I originally wrote them around the same time). But if you have a generic version we can deduplicate it all.

Reply:
I have my own version of this for my VF2 branch, and I _might_ be able to generalise this `Target` version of it to cover the case, but I can't remember entirely.

Reply:
I think let's just merge as is from the current PR head, and any changes we make in the future can be done as part of subsequent PRs anyway.